clients/da-vinci-client/src/main/java/com/linkedin/davinci/storage/StorageService.java

```diff
@@ -371,6 +375,35 @@ public synchronized AbstractStorageEngine openStore(
     return engine;
   }
 
+  public synchronized void checkWhetherStoragePartitionsShouldBeKeptOrNot(SafeHelixManager manager) {
+    if (manager == null) {
+      return;
+    }
+    for (AbstractStorageEngine storageEngine: getStorageEngineRepository().getAllLocalStorageEngines()) {
+      String storeName = storageEngine.getStoreVersionName();
+      Set<Integer> storageEnginePartitionIds = new HashSet<>(storageEngine.getPartitionIds());
+      String instanceHostName = manager.getInstanceName();
+      PropertyKey.Builder propertyKeyBuilder =
+          new PropertyKey.Builder(configLoader.getVeniceClusterConfig().getClusterName());
+      SafeHelixDataAccessor helixDataAccessor = manager.getHelixDataAccessor();
+      IdealState idealState = helixDataAccessor.getProperty(propertyKeyBuilder.idealStates(storeName));
+
+      if (idealState != null) {
+        Map<String, Map<String, String>> mapFields = idealState.getRecord().getMapFields();
+        for (Integer partitionId: storageEnginePartitionIds) {
+          if (storageEngine.isMetadataPartition(partitionId)) {
+            continue;
+          }
+          String partitionDbName = storeName + "_" + partitionId;
+          if (!mapFields.containsKey(partitionDbName) || (!mapFields.get(partitionDbName).containsKey(instanceHostName)
+              && !mapFields.get(partitionDbName).isEmpty())) {
```

Comment:
Although `!mapFields.get(partitionDbName).isEmpty()` may make the tests pass, this isn't exactly desirable behavior.

I think your existing logic is correct, but there is something wrong with the tests failing. The instance should be assigned to the partition, but it's not. I will try to back to this later this week.

Reply:
For one case, ensuring the list of hosts is not empty and the current host is in the list of hosts for the partition is one case in which we delete the partition.

Reply:
It's possible for the `mapFields` to be empty for a `partitionDbName`, if any hosts haven't been assigned to it yet. However, in the tests that are failing, I can see that the host is being assigned in the ExternalView but not in the IdealState. So there's a problem with Helix not being updated in-time when we're trying to perform the cleanup.

Reply:
Right, I understand this possibility.  
I initially thought of one possibility that coincided with the check of current host being assigned that may have made sense.

Reply:
Commit [5e43f6f](https://github.com/linkedin/venice/commit/5e43f6f3671598deefbc6c4c3f8dea10ab0e9190) covers the change that causes the tests to pass (a check that initially could coincide). 
I have reverted the change.